### PR TITLE
LightmapGI: Pack L1 SH coefficients for directional lightmaps

### PR DESCRIPTION
**Pull request by BlueCube3310**
Depends on #96113

Packs the L1 coefficients of SH lightmaps into a 0.0-1.0 range, which will allow compressing directional lightmaps as BC6U. This is an approach based on https://media.contentapi.ea.com/content/dam/eacom/frostbite/files/gdc2018-precomputedgiobalilluminationinfrostbite.pdf.

The packing is currently performed as an extra step after the image has already been denoised as a way to avoid precision errors. 

TODO:
- [x] Add a testing project,
- [x] Compare the visual results between unpacked and packed,
- [x] Add a versioning system to LightmapGIData so as to not load old lightmaps and ask the user to rebake.

Future work:
- Fix overdarkening caused by the colors becoming negative (https://web.archive.org/web/20160313132301/http://www.geomerics.com/wp-content/uploads/2015/08/CEDEC_Geomerics_ReconstructingDiffuseLighting1.pdf),
- Consider saving L1 as LDR images. This will increase their precision, but the size will remain the same as they will require BC7 compression due to etcpak's BC1 being too low-quality,

Comparison:
| bake mode | 4.4-dev1 | PR |
|-------------|----------|----|
| Static | 44dev-static | pr-static |
| Dynamic | 44dev-dynamic | pr-dynamic |

Testing project:
[sh-lightmaps.zip](https://github.com/user-attachments/files/16761040/sh-lightmaps.zip)

### CODE REVIEW

**Comment by BlueCube3310** [comment]
I've added a basic versioning system with the use of a new internal property whose purpose is to notify the LightmapGI node if the directional lightmaps are packed or not. I'm not sure if this is a good solution though.

File: modules/lightmapper_rd/lm_compute.glsl [hunk 1]
```diff
@@ -1037,4 +1038,28 @@ void main() {
...
 #endif
+
+#ifdef MODE_PACK_L1_COEFFS
+	vec4 base_coeff = texelFetch(sampler2DArray(source_light, linear_sampler), ivec3(atlas_pos, params.atlas_slice * 4), 0);
+
+	for (int i = 1; i < 4; i++) {
+		vec4 c = texelFetch(sampler2DArray(source_light, linear_sampler), ivec3(atlas_pos, params.atlas_slice * 4 + i), 0);
+
+		if (abs(base_coeff.r) > 0.0) {
+			c.r /= (base_coeff.r * 8);
```
Comment (clayjohn) [inline modules/lightmapper_rd/lm_compute.glsl:1049]: Where is the 8 coming from? From the presentation I remember that they factored out the 2 and then post-multiplied that in the shader. I understand we are combining the [-1,1] -> [0,1] into one step, but then it should be 4 here and 2 in the reading shader. 

We are dividing and multiplying by an extra factor of 2 here as far as I can tell

File: scene/3d/lightmap_gi.cpp [hunk 2]
```diff
@@ -1352,6 +1365,7 @@ void LightmapGI::_notification(int p_what) {
 	switch (p_what) {
 		case NOTIFICATION_POST_ENTER_TREE: {
 			if (light_data.is_valid()) {
+				ERR_FAIL_COND_MSG(light_data->is_using_spherical_harmonics() && !light_data->is_using_packed_directional(), "The directional lightmap textures are stored in a format that isn't supported anymore. Please rebake to fix the issue.");
```
Comment (clayjohn) [inline scene/3d/lightmap_gi.cpp:1368]: Since the packing is implemented as a separate pass, could we not just run that pass when we encounter an old lightmap and upgrade it automatically for the user?

**Comment by clayjohn** [comment]
> * Fix overdarkening caused by the colors becoming negative (https://web.archive.org/web/20160313132301/http://www.geomerics.com/wp-content/uploads/2015/08/CEDEC_Geomerics_ReconstructingDiffuseLighting1.pdf),

Here is an MIT licenced implementation of the Geomerics technique. Its a but more expensive to unpack, but it looks really good. 

https://github.com/kayru/Probulator/blob/4a97a2b021eb2ca7ef696f4ddf36ba9a9432cbb6/Source/Probulator/SphericalHarmonics.h#L136-L151

> * Consider saving L1 as LDR images. This will increase their precision, but the size will remain the same as they will require BC7 compression due to etcpak's BC1 being too low-quality,

We can probably switch to BC1 once we merge https://github.com/godotengine/godot/pull/95915

File: modules/lightmapper_rd/lm_compute.glsl (hunk 1 above)
Comment (BlueCube3310) [reply modules/lightmapper_rd/lm_compute.glsl:1049]: In the paper the L1 data is also multiplied by 0.5 at bake time. Godot doesn't do that, so we have to include the extra '2' in the division.

File: scene/3d/lightmap_gi.cpp (hunk 2 above)
Comment (BlueCube3310) [reply scene/3d/lightmap_gi.cpp:1368]: I thought about it, but older lightmaps wouldn't work the same way since they use different coefficients

Comment (clayjohn) [reply scene/3d/lightmap_gi.cpp:1368]: Oh right, I guess https://github.com/godotengine/godot/pull/95888 technically breaks compatibility already for existing lightmaps

File: modules/lightmapper_rd/lm_compute.glsl (hunk 1 above)
Comment (BlueCube3310) [reply modules/lightmapper_rd/lm_compute.glsl:1049]: In the paper, this is mentioned under the diffuse evaluation:
aa

I might be misunderstanding this, but in the paper the L1 data is 2x as big as L0, but it needs to be smaller or equal, hence the first division. Then it's in the [-1 - 1] range, so it's divided again and 0.5 is added to place it in the [0 - 1] range. In summary, the total value they need to be divided by is 4.

 In Godot, the L1 harmonics are 4x bigger than L0, which needs to be reflected when packing. The first division by 4 brings the data into the [-1 - 1] range, and another by 2 (and +0.5) into [0 - 1], hence the 8

Comment (clayjohn) [reply modules/lightmapper_rd/lm_compute.glsl:1049]: We use the same coefficients as the paper.

In the paper they multiply by two outside of the "EvaluateDiffuseL1()" function. In Godot I factored out the multiplication by 2 and the multiplication by 0.5 which is why our L1 coefficients are seemingly multiplied by nothing.

Comment (Zireael07) [reply modules/lightmapper_rd/lm_compute.glsl:1049]: Whatever you end up doing please document this sort of stuff for future contributors so that people don't end up throwing things off when trying to optimize

Comment (clayjohn) [reply modules/lightmapper_rd/lm_compute.glsl:1049]: > Whatever you end up doing please document this sort of stuff for future contributors so that people don't end up throwing things off when trying to optimize

This is the second time you have suggested this despite the code already having extensive comments: https://github.com/godotengine/godot/blob/88197d4a513a7873cb7cf85138a42a04fb1c9011/modules/lightmapper_rd/lm_compute.glsl#L652-L655

The problem is this stuff is super confusing and math heavy. So even with a detailed comment it hard to get things right. As you can see above both BlueCube and I both misunderstood aspects of how the SH should be encoded despite having comments

Comment (Zireael07) [reply modules/lightmapper_rd/lm_compute.glsl:1049]: Github commit view can get weird in large PRs. Totally missed some of the comments that already exist :/

Comment (clayjohn) [reply modules/lightmapper_rd/lm_compute.glsl:1049]: > edit: as I go through the presentation again, I see that the coefficients we use are wrong. The 2 times multiplier is what is left after multiplying the 0.5 coefficient by the 4 * PI from the Monte Carlo estimation. Similarly L0 should be multiplied by 1 (the PI is factored out by the BRDF).

On second thought, the results look fine now. Also, for direct lighting we don't do a Monte Carlo estimation. It seems we also already bake in the `1/PI` from the BRDF. So my math is probably totally wrong but somehow arriving at something similar to the correct values. I have a feeling that we should be integrating over a hemisphere rather than a sphere. I'll check

**Review by Calinou — APPROVED** [review]
Tested locally (rebased on top of `master` b6223c0df0300ba2db17b5742c349f13c33f8884), it works as expected in all rendering methods. This allows using VRAM compressed directional lightmaps with any rendering method, whereas you'd get error spam previously (and a black screen).

**Testing project:** [test_lightmap_directional_compress.zip](https://github.com/user-attachments/files/16882023/test_lightmap_directional_compress.zip)

## Preview

### VRAM Uncompressed

VRAM Uncompressed

### VRAM Compressed *(this PR)*

VRAM Compressed

PS: I noticed directional lightmaps generally represent light direction in a more faint manner compared to real-time lights, even when there is only a single clear light direction at a given texel, no indirect light involved and when lightmap bounces are set to 0. Is there a way we could get the directional lightmapping appearance to make normal mapped surfaces closer to how they'd look like with real-time lighting? This isn't an issue specific to this PR, but I thought it'd be worth pointing out.

File: scene/3d/lightmap_gi.cpp (hunk 2 above)
Comment (Calinou) [inline scene/3d/lightmap_gi.cpp:1368]: Display resource path and node name to make troubleshooting eassier, and reword the message to describe more clearly what happens in practice:

```suggestion
				ERR_FAIL_COND_MSG(light_data->is_using_spherical_harmonics() && !light_data->is_using_packed_directional(), vformat("%s (%s): The directional lightmap textures are stored in a format that isn't supported anymore. Please bake lightmaps again to make lightmaps display from this node again.", get_light_data()->get_path(), get_name()));
```

**Comment by clayjohn** [comment]
> PS: I noticed directional lightmaps generally represent light direction in a more faint manner compared to real-time lights, even when there is only a single clear light direction at a given texel, no indirect light involved and when lightmap bounces are set to 0. Is there a way we could get the directional lightmapping appearance to make normal mapped surfaces closer to how they'd look like with real-time lighting? This isn't an issue specific to this PR, but I thought it'd be worth pointing out.

I have been thinking about this a lot. I think it has to do with our discussion above about the coefficients. Right now I think something is off with the spherical harmonic baking process that results in the average light being roughly correct, but the directional light being too weak.

File: modules/lightmapper_rd/lm_compute.glsl (hunk 1 above)
Comment (clayjohn) [reply modules/lightmapper_rd/lm_compute.glsl:1049]: I'm making some progress on this! Indeed the 4 * PR factor should not be used for the direct pass. For the indirect pass, it should be 2 * PI, but we don't weight the samples, so no normalization is needed there. 

However, https://www.ppsloan.org/publications/StupidSH36.pdf (at page 12) indicates that the normalization factor for Directional Lights is `16 * PI / 17`. I'm about to try recalculating the coefficients but replacing `4 * PI` with `16 * PI / 17` and removing the BRDF's `1 / PR` completely

Comment (jcostello) [reply modules/lightmapper_rd/lm_compute.glsl:1049]: > Also, as I research this, I am finding that the way we bake static lighting for omnilights and spotlights is just completely wrong. We are essentially approximating them as if they were a directionalight.

@clayjohn that will change the way they look?

Comment (clayjohn) [reply modules/lightmapper_rd/lm_compute.glsl:1049]: > > Also, as I research this, I am finding that the way we bake static lighting for omnilights and spotlights is just completely wrong. We are essentially approximating them as if they were a directionalight.
> 
> @clayjohn that will change the way they look?

Yes, but I think the difference will be quite subtle. Basically they will appear softer than they should which is probably fine in most cases. I'm not going to worry about it unless someone tells me their Omni/Spot lights look horrible

Comment (clayjohn) [reply modules/lightmapper_rd/lm_compute.glsl:1049]: @BlueCube3310 I've discovered enough that I realize its probably best to maintain the division by 8 here and just live with this small hack for now. 

I am now totally certain that we are baking the spherical harmonics for direct lighting wrong. However, things look fine in most cases, so I am happy to live with the problems until we fix things properly. it doesn't make sense for this PR to wait until Directional baking is fully fixed. 

In other words, I think this is the correct code given the problems that exist for the directional baking right now.

**Review by clayjohn — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!
